### ISSUE TITLE
"Thread was being aborted" when using async keyword

### ISSUE THREAD

**Issue by AmadeusW** [octokit/octokit.net#853]
* What you were trying to do.
  Run the sample code, make simple API request

* What you expected to happen.
  See number of followers of the specified user, or a meaningful exception

* What actually happened.
  Got "Thread was being aborted." exception

* Steps to reproduce the problem.

  Run this method:
```csharp
        public async Task<string> CreateGist(string contents)
        {
            try
            {
                var github = new GitHubClient(new ProductHeaderValue("MyAmazingApp"));
                var user = await github.User.Get("half-ogre");
                Console.WriteLine(user.Followers + " folks love the half ogre!");
            }
            catch (Exception ex)
            {
                var x = ex;
            }
            return String.Empty;
        }
```
this exception will be caught while executing `var user = await github.User.Get("half-ogre");`
```
ex
{"Thread was being aborted."}
    Data: {System.Collections.ListDictionaryInternal}
    ExceptionState: Unable to evaluate expression because the code is optimized or a native frame is on top of the call stack.
    HResult: -2146233040
    HelpLink: null
    InnerException: null
    Message: "Thread was being aborted."
    Source: "System"
    StackTrace: "   at System.Net.HttpWebRequest.get_UseCustomHost()\r\n   at System.Net.HttpWebRequest.UpdateHeaders()\r\n   at System.Net.HttpWebRequest.SubmitRequest(ServicePoint servicePoint)\r\n   at System.Net.HttpWebRequest.BeginGetResponse(AsyncCallback callback, Object state)\r\n   at System.Net.Http.HttpClientHandler.StartGettingResponse(RequestState state)\r\n   at System.Net.Http.HttpClientHandler.StartRequest(Object obj)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Octokit.Internal.RedirectHandler.<SendAsync>d__0.MoveNext()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Octokit.Internal.HttpClientAdapter.<Send>d__1.MoveNext()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task t
ask)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Octokit.Connection.<RunRequest>d__2b.MoveNext()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Octokit.Connection.<Run>d__27`1.MoveNext()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Octokit.ApiConnection.<Get>d__0`1.MoveNext()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()\r\n   at gitIntegration.GitIntegration.<CreateGist>d__0.MoveNext() in C:\\Users\\Amadeus\\Documents\\GitHub\\SnippetVS\\gitIn
tegration\\gitIntegration\\GitIntegration.cs:line 17"
    TargetSite: {Boolean get_UseCustomHost()}
```

**Comment by AmadeusW** [comment]
The code works when I use `var user = github.User.Get("half-ogre").Result;`, but this breaks the await-async pattern.

**Comment by shiftkey** [comment]
@AmadeusW interesting exception - can you tell me whether you're running this from a console application or within an ASP.NET app?

**Comment by AmadeusW** [comment]
I was running it from Visual Studio extension in VS 2015 (RTM).

To get my request to work, I started a new process from within the extension, which was a console app. It worked there.

Would using an appdomain help in this case?

**Comment by shiftkey** [comment]
Interesting.

@shana have you seen this before? any tips?

**Comment by Haacked** [comment]
Try calling `ConfigureAwait(false)` like this:

`var user = await github.User.Get("half-ogre").ConfigureAwait(false);`

**Comment by AmadeusW** [comment]
I've created a unit test that executes the same Octokit code as the Visual Studio extension.

All of `var user = await github.User.Get("half-ogre");`, `var user = github.User.Get("half-ogre").Result;`  and `var user = await github.User.Get("half-ogre").ConfigureAwait(false);` work from a unit test context, but neither works from the Visual Studio extension.

I will port this to VS2013 and test it there

**Comment by AmadeusW** [comment]
Visual Studio 2013 is also frozen. For the last few days I can't reproduce the exception - code just hangs.

**Comment by Haacked** [comment]
Hmm, it's unclear to me why this would not work in VS2013 but work in your unit test context. Have you run a Fiddler trace to make sure the request is going out?

**Comment by shana** [comment]
I just ran your sample code from the onclick handler of a ToolWindow:

```
        async void button1_Click(object sender, RoutedEventArgs e)
        {
            await Connect();
        }

        async Task<string> Connect()
        {
            try {
                GitHubClient client = new GitHubClient(new ProductHeaderValue("my.test.app"));
                var user = await client.User.Get("half-ogre");
                MessageBox.Show(string.Format(CultureInfo.CurrentUICulture, "Invoked '{0}'", user.Followers), "ToolWindow1");
            }
            catch (Exception ex)
            {
                MessageBox.Show(string.Format(CultureInfo.CurrentUICulture, "Exception '{0}'", ex), "ToolWindow1");
            }
            return string.Empty;
        }
```

It worked fine, showing a messagebox with "Invoked '75'".

Your stacktrace suggests the calling thread is aborted just as the request is being built, which means it's not a timeout on the http side, but the environment aborting the thread. It might be that the way you call your CreateGist method is causing the thread to abort early, for some reason. Are you invoking it from the main thread, or from another thread?

**Comment by AmadeusW** [comment]
Thank you all for help!
@Haacked fiddler has shown that the request has completed and GitHub returned a response.
This helped me detect a [deadlock I created by misunderstanding of the async pattern](http://blog.ciber.no/2014/05/19/using-task-configureawaitfalse-to-prevent-deadlocks-in-async-code/). 
The exception I posted at the beginning of the thread didn't seem to come to play and just confused me.
